File: Sorting/topological-sort.java
// Topological Sort
// Type: DAG, DFS-based
// Time: O(V + E)
// Space: O(V)

import java.util.*;

public class TopoSort {
    public static List<Integer> topologicalSort(Map<Integer, List<Integer>> graph) {
        Set<Integer> visited = new HashSet<>();
        Stack<Integer> stack = new Stack<>();

        for (int node : graph.keySet()) {
            if (!visited.contains(node)) {
                dfs(node, visited, stack, graph);
            }
        }

        List<Integer> result = new ArrayList<>();
        while (!stack.isEmpty()) result.add(stack.pop());
        return result;
    }

    private static void dfs(int node, Set<Integer> visited, Stack<Integer> stack, Map<Integer, List<Integer>> graph) {
        visited.add(node);
        for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
            if (!visited.contains(neighbor)) dfs(neighbor, visited, stack, graph);
        }
        stack.push(node);
    }
}

File: Sorting/topological-sort.c
// Topological Sort
// Type: DAG, DFS-based
// Time: O(V + E)
// Space: O(V)

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX 100

int graph[MAX][MAX], visited[MAX], stack[MAX], top = -1, V;

void dfs(int node) {
    visited[node] = 1;
    for (int i = 0; i < V; i++) {
        if (graph[node][i] && !visited[i]) {
            dfs(i);
        }
    }
    stack[++top] = node;
}

void topoSort() {
    for (int i = 0; i < V; i++) {
        if (!visited[i]) dfs(i);
    }
    while (top >= 0) {
        printf("%d ", stack[top--]);
    }
}
